packages/vite/src/node/build.ts

```diff
@@ -541,16 +541,59 @@ export async function build(
     },
   }
 
+  /**
+   * The stack string usually contains a copy of the message at the start of the stack.
+   * If the stack starts with the message, we remove it and just return the stack trace
+   * portion. Otherwise the original stack trace is used.
+   */
+  function extractStack(e: RollupError) {
+    const { stack, name = 'Error', message } = e
+
+    // If we don't have a stack, not much we can do.
+    if (!stack) {
+      return stack
+    }
+
+    const expectedPrefix = `${name}: ${message}\n`
+    if (stack.startsWith(expectedPrefix)) {
+      return stack.slice(expectedPrefix.length)
+    }
+
+    return stack
+  }
+
+  /**
+   * Esbuild code frames have newlines at the start and end of the frame, rollup doesn't
+   * This function normalizes the frame to match the esbuild format which has more pleasing padding
+   */
+  const normalizeCodeFrame = (frame: string) => {
+    const trimmedPadding = frame.replace(/^\n|\n$/g, '')
+    return `\n${trimmedPadding}\n`
+  }
+
   const mergeRollupError = (e: RollupError) => {
```

Comment:
Thanks for the extensive explanation of the problem this is fixing. I think this PR makes sense, but it seems that now `mergeRollupError` is mutating the error, which could be surprising down the line.

From what I can tell, we can't avoid the mutation, but maybe it's better to rename this as `enhanceRollupError` to better reflect what it's doing. And it doesn't need to return anything as it's already mutating the error. Then we can update the consumers to take account of this change.